File: haskell/005/euler.hs
import Data.List

main :: IO ()

findPrimeFactors :: (Integral a) => a -> a -> [a]
findPrimeFactors n p
    | n < (p * p)  = [n]
    | mod n p == 0 = [p] ++ (findPrimeFactors (div n p) p)
    | otherwise    = findPrimeFactors n (p + 1)

primeFactors :: (Integral a) => a -> [a]
primeFactors n = findPrimeFactors n 2

overlap :: (Ord a) => [a] -> [a] -> [a]
overlap (x:xs) []     = (x:xs)
overlap [] (y:ys)     = (y:ys)
overlap (x:xs) (y:ys) = sort ((x:xs) ++ (o (x:xs) (y:ys)))
                       where
                         o (x:xs) []      = (x:xs)
                         o [] (y:ys)      = (y:ys)
                         o (x:xs) (y:ys)  = o xs (delete x (y:ys))

main = print(foldl' (*) 1 (foldr overlap [] (map primeFactors [1..20])))

File: nim/005/euler.nim
from sequtils import `concat`, `foldr`

proc prime_factors(n: int64): seq[int64] =
  var found = false

  var v = 2

  var i = 2

  while i < n - 1 and found == false:
    if n mod i == 0:
      found = true

      v = i

    i = i + 1

  result = @[]

  if found:
    add(result, v)

    result = concat(result, prime_factors(n div v))
  else:
    add(result, n)

  return

proc overlap(a: seq[int64], b: seq[int64]): seq[int64] =
  var c = a

  for i in b:
    let index = find(c, i)

    if index >= 0:
      delete(c, index)

  result = concat(c, b)

  return

var factors: seq[int64]

factors = @[]

for i in 1..20:
  factors = overlap(prime_factors(i), factors)

echo(foldr(factors, a * b))
